ptf/tests/linerate/meter_tests.py

```diff
@@ -0,0 +1,654 @@
+        pg_id_ue1 = 1
+        # UE2 above the session meter rate
+        stream_bps_ue2 = 100 * M
+        pg_id_ue2 = 2
+
+        streams = [
+            self.create_gtp_stream(
+                ue_addr=UE1_ADDR, teid=UE1_UL_TEID,
+                pg_id=pg_id_ue1, l1_bps=stream_bps_ue1, l2_size=1400
+            ),
+            self.create_gtp_stream(
+                ue_addr=UE2_ADDR, teid=UE2_UL_TEID,
+                pg_id=pg_id_ue2, l1_bps=stream_bps_ue2, l2_size=1400
+            )
+        ]
+
+        self.trex_client.add_streams(streams, ports=TREX_OUT_PORT)
+        print(f"Starting traffic, duration: {TRAFFIC_DURATION_SECONDS} sec")
+        self.trex_client.start(TREX_OUT_PORT, duration=TRAFFIC_DURATION_SECONDS)
+        live_stats = monitor_port_stats(self.trex_client)
+        self.trex_client.wait_on_traffic(ports=TREX_OUT_PORT, rx_delay_ms=100)
+        avg_live_stats = self.average_monitored_port_stats(live_stats)
+
+        # Get and print TREX stats
+        trex_stats = self.trex_client.get_stats()
+        lat_stats_ue1 = get_latency_stats(pg_id_ue1, trex_stats)
+        lat_stats_ue2 = get_latency_stats(pg_id_ue2, trex_stats)
+        flow_stats_ue1 = get_flow_stats(pg_id_ue1, trex_stats)
+        flow_stats_ue2 = get_flow_stats(pg_id_ue2, trex_stats)
+
+        rx_bps_ue1 = (flow_stats_ue1.rx_bytes * 8) / TRAFFIC_DURATION_SECONDS
+        rx_bps_ue2 = (flow_stats_ue2.rx_bytes * 8) / TRAFFIC_DURATION_SECONDS
+
+        print("============= STATS UE 1 =============")
+        print(f"   RX Mbps: {rx_bps_ue1/M}")
+        print(get_readable_flow_stats(flow_stats_ue1))
+        print(get_readable_latency_stats(lat_stats_ue1))
+        print(SEPARATOR)
+        print("============= STATS UE 2 =============")
+        print(f"   RX Mbps: {rx_bps_ue2/M}")
+        print(get_readable_flow_stats(flow_stats_ue2))
+        print(get_readable_latency_stats(lat_stats_ue2))
+        print(SEPARATOR)
+
+        self.assertEqual(
+            lat_stats_ue1.dropped,
+            0,
+            "Conforming traffic shouldn't get packet drops"
+        )
+        self.assertGreater(
+            lat_stats_ue2.dropped,
+            0,
+            "Misbehaving traffic should get packet drops"
+        )
+        self.assertAlmostEqual(
+            avg_live_stats["avg_rx"][TREX_IN_PORT] / slice_rate,
+            1,
+            delta=0.1,
+            msg="Traffic should be almost equal the slice rate"
+        )
+        self.assertAlmostEqual(
+            rx_bps_ue1 / stream_bps_ue1,
+            1,
+            delta=0.05,
+            msg="UE 1 traffic (same as session rate) should not be shaped"
+        )
+        self.assertAlmostEqual(
+            rx_bps_ue2 / session_rate,
+            1,
+            delta=0.05,
+            msg="UE 2 traffic (above session rate) should be shaped to session rate"
+        )
+
+
+@group("trex-sw-mode")
+@group("upf")
+@group("meter")
+class ColorAwareSessionMeter(UpfPolicingTest):
```

Comment:
Many of the previous comments apply to this class as well. I'll let you do a first pass before commenting further.